src/main/java/org/jenkinsci/plugins/workflow/graphanalysis/ForkScanner.java

```diff
@@ -553,11 +554,49 @@ public static void visitSimpleChunks(@Nonnull Collection<FlowNode> heads, @Nonnu
         scanner.visitSimpleChunks(visitor, finder);
     }
 
-    /** Walk through flows  */
+    /** Ensures we find the last *begun* node when there are multiple heads (parallel branches)
+     *  This means that the simpleBlockVisitor gets the *actual* last node, not just the end of the last declared branch
+     *  (See issue JENKINS-38536)
+     */
+    @CheckForNull
+    private static FlowNode findLastStartedNode(@Nonnull List<FlowNode> candidates) {
+        if (candidates.size() == 0) {
+            return null;
+        } else if (candidates.size() == 1) {
+            return candidates.get(0);
+        } else {
+            FlowNode returnOut = candidates.get(0);
+            long startTime = Long.MIN_VALUE;
+            for(FlowNode f : candidates) {
+                TimingAction ta = f.getAction(TimingAction.class);
+                // Null timing with multiple heads is probably a node where the GraphListener hasn't fired to add TimingAction yet
+                long myStart = (ta == null) ? System.currentTimeMillis() : ta.getStartTime();
```

Comment:
@oleg-nenashev Agreed, I'm not a huge fan of having to make this assumption either -- looking at hardening up TimingAction addition (other PR, and a bit more investigation) because it's been a pain point.

However we *can't* safely skip these nodes either, because if present they are likely to be the most recently created node (there's a tiny window where it *might* be possible to have a brand new node without the action, in which case it would be the newest one -- just about impossible to test though because it would require precise timing).   Note that it's quite possible (and likely) that the action would be attached between this code block and the next time we poke the FlowNode.

In any case: it's likely a rare/never situation barring the issue noted with Semaphore steps, and I can't think of a more robust way to handle it.